### PR TITLE
Opengl glext private

### PR DESCRIPTION
**Pull request by devel0**
## What does the pull request do?

Allow to write custom extension well prototyped, hiding entrypoint (private) property with a public method that forward to that.

## What is the current behavior?

Can't declare entrypoint as private, it must public and that makes 2 member visible when using intellisense ( property and method ).

## What is the updated/expected behavior with this PR?

With this I can write something like the follow to implement my own extension with method documentation:

```csharp
using System;
using System.Runtime.CompilerServices;
using Avalonia.OpenGL;

namespace SearchAThing.OpenGL
{

    public unsafe partial class GlInterface : GlInterfaceBase<Avalonia.OpenGL.GlInterface.GlContextInfo>
    {
        public GlInterface(Avalonia.OpenGL.GlInterface GL) : base(GL.GetProcAddress, GL.ContextInfo)
        {
        }
       
        delegate void _glActiveShaderProgram(uint pipeline, uint program);
        [GlEntryPoint("glActiveShaderProgram")]
        _glActiveShaderProgram _ActiveShaderProgram { get; }

        /// <summary>
        /// glActiveShaderProgram sets the linked program named by program            to be the active program for the program pipeline object pipeline. The active            program in the active program pipeline object is the target of calls to glUniform            when no program has been made current through a call to glUseProgram.
        /// </summary>
        /// <param name="pipeline">Specifies the program pipeline object to set the active program object for.</param>
        /// <param name="program">Specifies the program object to set as the active program pipeline object pipeline.</param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void ActiveShaderProgram(uint pipeline, uint program) => _ActiveShaderProgram(pipeline, program);

    }

}
```

and use them like [this example](https://github.com/devel0/netcore-sci/blob/1586e6ade5f41007573bf1b287f4a31f3493e328/examples/0006/Controls/OpenGlControl.cs#L329-L336)

```csharp
        SearchAThing.OpenGL.GlInterface GL;

        /// <summary>
        /// GL INIT
        /// </summary>
        protected unsafe override void OnOpenGlInit(Avalonia.OpenGL.GlInterface _GL, int fb)
        {
            GL = new SearchAThing.OpenGL.GlInterface(_GL);
        // ...
```

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

**Comment by devel0** [comment]
@kekekeks note that Avalonia.Native package may not used directly by the Linux platform, ok, but if not generated there will be errors during restore because some other packages still refers it as described in https://github.com/AvaloniaUI/Avalonia/issues/4148#issue-642261119 ; btw I can revert that commit if this will be addressed in future.

File: src/Avalonia.Native/Avalonia.Native.csproj [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <IsPackable>false</IsPackable>
-    <IsPackable Condition="'$([MSBuild]::IsOSPlatform(OSX))' == 'True'">true</IsPackable>
```
Comment (kekekeks) [inline src/Avalonia.Native/Avalonia.Native.csproj:5]: The check is here for a reason

File: src/Avalonia.OpenGL/GlInterfaceBase.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ public class GlInterfaceBase<TContext>
         public GlInterfaceBase(Func<string, IntPtr> getProcAddress, TContext context)
         {
             _getProcAddress = getProcAddress;
-            foreach (var prop in this.GetType().GetProperties())
+            foreach (var prop in this.GetType().GetProperties(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static))
```
Comment (kekekeks) [inline src/Avalonia.OpenGL/GlInterfaceBase.cs:26]: Initializing static properties from an instance constructor is usually a bad idea.

Comment (devel0) [reply src/Avalonia.OpenGL/GlInterfaceBase.cs:26]: There is other way to access to private members ?

File: src/Avalonia.Native/Avalonia.Native.csproj (hunk 1 above)
Comment (devel0) [reply src/Avalonia.Native/Avalonia.Native.csproj:5]: ok, but that not address the issue https://github.com/AvaloniaUI/Avalonia/issues/4148#issue-642261119

File: src/Avalonia.OpenGL/GlInterfaceBase.cs (hunk 2 above)
Comment (kekekeks) [reply src/Avalonia.OpenGL/GlInterfaceBase.cs:26]: Yes, but not including `BindingFlags.Static`. 
See https://docs.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=netcore-3.1#System_Type_GetProperties_System_Reflection_BindingFlags_ and https://docs.microsoft.com/en-us/dotnet/api/system.reflection.bindingflags?view=netcore-3.1

Comment (devel0) [reply src/Avalonia.OpenGL/GlInterfaceBase.cs:26]: You right, thanks for clarification.

**Comment by devel0** [comment]
@kekekeks I ended up using Silk.NET as api for GL so there is no need for me to add more api with private delegates.

Its not a clean solution like the OpenGLControlBase of Avalonia that uses framebuffer, btw everything worked as expected apart the fact that in X11 I got crashes if I close a modal window or if I use a tooltip ( may related to #3536 ), in general every time time that another window created while in Windows there is no issues.
```
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  4 (X_DestroyWindow)
  Resource id in failed request:  0x56001e2
  Serial number of failed request:  2625
  Current serial number in output stream:  2627
```

Note that if I not open the silkNet window ( by commenting [this](https://github.com/devel0/netcore-sci/blob/097f3978a3a95d96c1060f99b973eca1fdc94a17/examples/0002/MainWindow.xaml#L22) the problem not appears ), so there are many possibilities that this is caused by my mistake in interleaving windowing mode but hopefully this could be of help to you to check against some robustness of the Avalonia.X11

If you want to look at the example just checkout the source:
```sh
git clone https://github.com/devel0/netcore-sci.git
cd netcore-sci
git checkout 097f3978a3a95d96c1060f99b973eca1fdc94a17
dotnet build
dotnet run --project examples/0002
```

image

**Pull request closed without merging**